apiserver/facades/agent/uniter/networkinfo.go

```
@@ -422,20 +411,142 @@ func (n *NetworkInfo) NetworksForRelation(
+				err, "getting machine %q preferred private address", machine.MachineTag())}
+
+			// Remove this ID to prevent further processing.
+			spaceSet.Remove(corenetwork.AlphaSpaceId)
+		} else {
+			privateIPAddress = privateMachineAddress.Value
+		}
 	}
 
-	return machine.GetNetworkInfoForSpaces(set.NewStrings(spaces...)), nil
+	// Link-layer devices are set in a single transaction for all devices
+	// observed on the machine, so the first result will include them all.
+	var addresses []*state.Address
+	retryArg := n.retryFactory()
+	retryArg.Func = func() error {
+		var err error
+		addresses, err = machine.AllAddresses()
+		return err
+	}
+	retryArg.IsFatalError = func(err error) bool {
+		return err != nil
+	}
+	if err := retry.Call(retryArg); err != nil {
+		result := machineNetworkInfoResult{Error: errors.Annotate(err, "getting devices addresses")}
+		for _, id := range spaceSet.Values() {
+			if _, ok := results[id]; !ok {
+				results[id] = result
+			}
+		}
+		return results, nil
+	}
+
+	logger.Debugf("Looking for address from %v in spaces %v", addresses, spaceIDs)
+
+	var privateLinkLayerAddress *state.Address
+	for _, addr := range addresses {
+		subnet, err := addr.Subnet()
+		switch {
+		case errors.IsNotFound(err):
+			logger.Debugf("skipping %s: not linked to a known subnet (%v)", addr, err)
+
+			// For a space-less model, we will not have subnets populated,
+			// and will therefore not find a subnet for the address.
+			// Capture the link-layer information for machine private address
+			// so that we can return as much information as possible.
+			// TODO (manadart 2020-02-21): This will not be required once
+			// discovery (or population of subnets by other means) is
+			// introduced for the non-space IAAS providers (LXD, manual, etc).
+			if addr.Value() == privateIPAddress {
+				privateLinkLayerAddress = addr
+			}
+		case err != nil:
+			logger.Errorf("cannot get subnet for address %q - %q", addr, err)
```

That reads like a kind of error that should not be muted. Shouldn't we return the error instead of log+skip here?

---

No, because we fall back outside of this loop when dealing with space-less providers.